### ISSUE TITLE
Scraping Witness Lists

### ISSUE THREAD

**Issue by AmyMosley** [open-austin/influence-texas#39]
The advanced user version of this product will involve users running complex searches and will require more server space, which costs money. Users could be charged a fee for this service. Does this functionality need to be built into the project now?

**Comment by mscarey** [comment]
An update for people interested in this issue: I didn't have much success trying to parse the witness lists with regexes. (see my [notebook](https://github.com/open-austin/influence-texas/blob/master/notebooks/witness-lists.ipynb) and the [output of that process](https://github.com/open-austin/influence-texas/tree/master/data/witness-lists)). When you download the bills, be sure to get them from the ftp directory (ftp://ftp.legis.state.tx.us/bills/85R/witlistbill/). I believe the house and senate lists follow slightly different formats, so you might want to write separate parsers. Even within the same chamber of the legislature, there will be inconsistencies in formatting. I think the listings are not always consistent about what kind of information goes in the parentheses, and some fields are omitted on some lines. It's just messy data in that some witnesses will have filled out the sign-up form wrong, and their mistakes end up in the documents. There are Python libraries that do named entity recognition, so I suggest looking at those.

**Comment by mscarey** [comment]
[Updated.](https://github.com/open-austin/influence-texas/blob/master/notebooks/witness-lists.ipynb) The notebook should be easier to read and I think the output files are about 95% correct. The notebook still just uses regexes, which will never be perfect where names are involved.

**Comment by jpolache** [comment]
RE: Witness List issue. I was able to download the files from the SOT web site and run @mscarey’s script, That got me a HouseWitness.csv with 722 entries.  I ran a script against that to match the names in the FullText field against the FirstName and LastName fields.  That gave me a list of 108 mismatches. I ran some of those against the regex /\w*.\w+\,.+\(/ and seemed to get good results.  I'd like to try it in the script, but I'm not sure how to insert it.

**Comment by mscarey** [comment]
@jpolache Are you saying you're having trouble adding new lines to the jupyter notebook and rerunning it? I'm not clear on where you got stuck, but maybe you want the [Jupyter documentation](https://jupyter-notebook.readthedocs.io/en/latest/)?

**Comment by jpolache** [comment]
Not sure where in the code to insert my regex. Also, not sure how your regexs work. I usually figure out other people's code by stepping through it in a debugger (not that your code has bugs :) ).  I'm not good enough with python pdb to figure it out.

**Comment by mscarey** [comment]
Fair enough, the code is hard to understand. I just committed a new version of the notebook that hopefully will be a little clearer. I added docstrings for each function. I also changed a few of the functions to use strings rather than lists as inputs and outputs, so I hope that doesn't break any code you've already written.

**Comment by jpolache** [comment]
Thanks Matt.  I'm working through the new code version now.

**Comment by jpolache** [comment]
@mscarey Getting an error :(

import csv

def export(dir, witList):
    with open(dir,'w') as f:
        writer = csv.writer(f)
        writer.writerow(['FullText', 'Position', 'Bill', 'LastName', 'FirstName', 'Role', 'Organization', 'City', 'State'])
        writer.writerows(witList) # better just to include the FullText field.
    return None

houseDir = 'C:/Users/user/Documents/code3/venv/witness/HouseWitness1205.csv'
#houseDir = '../data/witness-lists/HouseWitness.csv'
export(houseDir, houseRows)
***********************************************************************************************
NameError                                 Traceback (most recent call last)
<ipython-input-5-7a28b052e639> in <module>()
     10 houseDir = 'C:/Users/user/Documents/code3/venv/witness/HouseWitness1205.csv'
     11 #houseDir = '../data/witness-lists/HouseWitness.csv'
---> 12 export(houseDir, houseRows)

NameError: name 'houseRows' is not defined

**Comment by mscarey** [comment]
@jpolache is it possible you only ran the code in that block, but you didn't run the code earlier in the notebook where a list is assigned to the variable "houseRows"?

**Comment by jpolache** [comment]
This time, I went to the menu and selected Kernel->Restart and Run All.  Here is the error;

---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-6-4c041a472210> in <module>()
     22 folderName = 'C:/Users/user/Documents/code3/venv/witness/house_bills/HB00400_HB00499/'
     23 #folderName = 'bills/85R/witlistbill/html/house_bills/'
---> 24 houseWit = extractRows(folderName)

<ipython-input-6-4c041a472210> in extractRows(folderName)
     15                 wit = HBWitness(source)
     16                 # trying to rejoin entries split across lines
---> 17                 new = mergelines(wit)
     18                 new = mergelines(new) # Will doing it twice catch 3-line entries?
     19                 houseWit.extend(new)

<ipython-input-5-0ff9b8be6b9f> in mergelines(wit)
     28             changed += 1
     29 
---> 30         elif row[0].count(')') != row[0].count('(')         and wit[lineIndex + 1][0].count(')') != wit[lineIndex + 1][0].count('(')         and addName(row[0]) != [None, None] and not re.search(endWithState, row[0]):
     31             newList.append([row[0] + " " + wit[lineIndex + 1][0], row[1], row[2]])
     32             badList.append(wit[lineIndex + 1][0:3])

<ipython-input-3-a39f531de21f> in addName(line)
     17     for f in flags:
     18         for r in regexes:
---> 19             nameRe = re.compile(r, f)
     20             match = re.search(nameRe, line)
     21             if match:

c:\users\user\documents\code3\venv\lib\re.py in compile(pattern, flags)
    231 def compile(pattern, flags=0):
    232     "Compile a regular expression pattern, returning a pattern object."
--> 233     return _compile(pattern, flags)
    234 
    235 def purge():

c:\users\user\documents\code3\venv\lib\re.py in _compile(pattern, flags)
    299     if not sre_compile.isstring(pattern):
    300         raise TypeError("first argument must be string or compiled pattern")
--> 301     p = sre_compile.compile(pattern, flags)
    302     if not (flags & DEBUG):
    303         if len(_cache) >= _MAXCACHE:

c:\users\user\documents\code3\venv\lib\sre_compile.py in compile(p, flags)
    560     if isstring(p):
    561         pattern = p
--> 562         p = sre_parse.parse(p, flags)
    563     else:
    564         pattern = None

c:\users\user\documents\code3\venv\lib\sre_parse.py in parse(str, flags, pattern)
    853 
    854     try:
--> 855         p = _parse_sub(source, pattern, flags & SRE_FLAG_VERBOSE, 0)
    856     except Verbose:
    857         # the VERBOSE flag was switched on inside the pattern.  to be

c:\users\user\documents\code3\venv\lib\sre_parse.py in _parse_sub(source, state, verbose, nested)
    414     while True:
    415         itemsappend(_parse(source, state, verbose, nested + 1,
--> 416                            not nested and not items))
    417         if not sourcematch("|"):
    418             break

c:\users\user\documents\code3\venv\lib\sre_parse.py in _parse(source, state, verbose, nested, first)
    766             if not source.match(")"):
    767                 raise source.error("missing ), unterminated subpattern",
--> 768                                    source.tell() - start)
    769             if group is not None:
    770                 state.closegroup(group, p)

error: missing ), unterminated subpattern at position 10

**Comment by jpolache** [comment]
Fixed the "error: missing ), unterminated subpattern at position 10" issue by using an escape "\(" on the open peren in my regex pattern.  But the regex pattern is not compatible with the rest of the script, apparently because I am not using grouping (?:). So far I am not able to duplicate the regex I came up with using grouping and am hesitant to try and rewrite everything else to resolve the error.  Research continues.

**Comment by mscarey** [comment]
@jpolache I can't quite tell which regular expression is triggering the error for you. Maybe the issue is that I was running a different version of Python or a library. I added lines to the [notebook](https://github.com/open-austin/influence-texas/blob/master/notebooks/witness-lists.ipynb) that show what I was running. Here's what they show:

```
Python 3.6.1 :: Continuum Analytics, Inc.
bs4 4.6.0
pandas 0.20.3
```

**Comment by jpolache** [comment]
Thanks Matt. I'll check it out.

**Comment by jpolache** [comment]
Matt,

Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 18:11:49) [MSC v.1900 64 bit (AMD64)]

beautifulsoup4==4.6.0
bs4==0.0.1
pandas==0.21.0

[freeze.txt](https://github.com/open-austin/influence-texas/files/1561598/freeze.txt)

**Comment by lazarus1331** [comment]
I will be at the meetup on Monday evening.  Will Matt or John be there as
well?  If so, we should be able to easily resolve this.

-Michael

**Comment by mscarey** [comment]
@lazarus1331 Yeah, I'll be there. There isn't usually a bunch of time to do project work at the library, but I'll bring my computer.

**Comment by jpolache** [comment]
Sorry I was unable to attend the meetup.  I now have the code running in my
environment and have done some analysis of the output.

Let me know if you would like to discuss next steps.

Jonathan
512 659 6919
